react/features/local-recording/recording/WavAdapter.js

```diff
@@ -0,0 +1,284 @@
+                this._wavLength = 0;
+                this._wavBuffers.push(this._createWavHeader());
+
+                this._audioSource.connect(this._audioProcessingNode);
+                this._audioProcessingNode
+                    .connect(this._audioContext.destination);
+                resolve();
+            });
+    }
+
+    /**
+     * Implements {@link RecordingAdapter#stop()}.
+     *
+     * @inheritdoc
+     */
+    stop() {
+        this._audioProcessingNode.disconnect();
+        this._audioSource.disconnect();
+        this._data = this._exportMonoWAV(this._wavBuffers, this._wavLength);
+
+        return Promise.resolve();
+    }
+
+    /**
+     * Implements {@link RecordingAdapter#download()}.
+     *
+     * @inheritdoc
+     */
+    download() {
+        if (this._data !== null) {
+            const audioURL = window.URL.createObjectURL(this._data);
+
+            downloadBlob(audioURL, `recording${timestampString()}.wav`);
+        }
+
+    }
+
+    /**
+     * Creates a WAVE file header.
+     *
+     * @private
+     * @returns {Uint8Array}
+     */
+    _createWavHeader() {
+        // adapted from
+        // https://github.com/mmig/speech-to-flac/blob/master/encoder.js
+
+        // ref: http://soundfile.sapp.org/doc/WaveFormat/
+
+        // create our WAVE file header
+        const buffer = new ArrayBuffer(44);
+        const view = new DataView(buffer);
+
+        // RIFF chunk descriptor
+        writeUTFBytes(view, 0, 'RIFF');
+
+        // set file size at the end
+        writeUTFBytes(view, 8, 'WAVE');
+
+        // FMT sub-chunk
+        writeUTFBytes(view, 12, 'fmt ');
+        view.setUint32(16, 16, true);
+        view.setUint16(20, 1, true);
+
+        // NumChannels
+        view.setUint16(22, 1, true);
+
+        // SampleRate
+        view.setUint32(24, WAV_SAMPLE_RATE, true);
+
+        // ByteRate
+        view.setUint32(28,
+            Number(WAV_SAMPLE_RATE) * 1 * WAV_BITS_PER_SAMPLE / 8, true);
+
+        // BlockAlign
+        view.setUint16(32, 1 * Number(WAV_BITS_PER_SAMPLE) / 8, true);
+
+        view.setUint16(34, WAV_BITS_PER_SAMPLE, true);
+
+        // data sub-chunk
+        writeUTFBytes(view, 36, 'data');
+
+        // DUMMY file length (set real value on export)
+        view.setUint32(4, 10, true);
+
+        // DUMMY data chunk length (set real value on export)
+        view.setUint32(40, 10, true);
+
+        return new Uint8Array(buffer);
+    }
+
+
+    /**
+     * Callback function that saves the PCM bits.
+     *
+     * @private
+     * @param {Float32Array} data - The audio PCM data.
+     * @returns {void}
+     */
+    _saveWavPCM(data) {
+        // need to copy the Float32Array,
+        // unlike passing to WebWorker,
+        // this data is passed by reference,
+        // so we need to copy it, otherwise the
+        // audio file will be just repeating the last
+        // segment.
+        this._wavBuffers.push(new Float32Array(data));
```

Comment:
Is there still any size concern with saving the buffers in memory? What happens in a long call, like an hour call?

Reply:
Yeah, that is the next step, saving the parts of the buffer in IndexedDB. But practically, with the current codebase I have tested recording a 90min call and it seems to be working fine (the resulting flac file is ~95MB).